client/app/bundles/course/reference-timelines/components/TimelinesStack/TimelinesStack.tsx

```diff
@@ -0,0 +1,70 @@
+import { Fragment } from 'react';
+import {
+  ItemWithTimeData,
+  TimeData,
+  TimelineData,
+} from 'types/course/referenceTimelines';
+
+import { useAppSelector } from '../../store';
+import RowSpacer from '../RowSpacer';
+
+import AssignableTimeline from './AssignableTimeline';
+import AssignedTimeline from './AssignedTimeline';
+
+/**
+ * Maximum inner width of the calendar view. It is essentially equals to
+ * `MAX_DAYS` times `DAY_WIDTH_PIXELS`.
+ */
+const MAX_CALENDAR_INNER_WIDTH_PIXELS = '3e+09px' as const;
+
+interface TimeBarsProps {
+  for: ItemWithTimeData[];
+  within: TimelineData[];
+}
+
+const TimelinesStack = (props: TimeBarsProps): JSX.Element => {
+  const { for: items, within: timelines } = props;
+  const gamified = useAppSelector((state) => state.timelines.gamified);
+
+  return (
+    <div
+      className="relative space-y-[1px]"
+      style={{ width: MAX_CALENDAR_INNER_WIDTH_PIXELS }}
+    >
+      {items.map((item) => {
+        let defaultTime: TimeData;
+
+        return (
+          <Fragment key={item.id}>
+            {timelines.map((timeline) => {
+              const time = item.times[timeline.id];
+              if (timeline.default) defaultTime = time;
+
+              return time ? (
+                <AssignedTimeline
+                  key={timeline.id}
+                  for={item}
+                  gamified={gamified}
+                  in={timeline}
+                  visualising={time}
+                />
+              ) : (
+                <AssignableTimeline
+                  key={timeline.id}
+                  basedOn={defaultTime}
```

Comment:
Did you actually manage to get an empty form?

This loop works on the invariant that a default timeline will always have a time. Therefore, it is impossible for a default timeline to be rendered as `AssignableTimeline` by `TimelinesStack`. In another words, if we define that there are exhaustively two types of timelines based on their times: assigned and assignable, a default timeline is always an assigned timeline.

Reply:
Okay I know why. A combination of line 35: `let defaultTime: TimeData;` and Line 41: `if (timeline.default) defaultTime = time` assumes that the first timeline is always the default one. Typescript does not scream since it assumes that defaultTime is somewhat already defined when it is typed with `TimeData` albeit it still being empty.

This might not be the case if we set up the weight properly. But we dont want to assume that. So I would suggest that the `defaultTime` line 35 should find the time of the default time.

Reply:
Oh, I get what you mean. Yes, I missed that implicit precondition in my previous message. I am too not sure why TypeScript does not automatically add `| undefined` to undefined `let`s.

Your suggestion would solve the issue, but that will add an O(n) search for every item. I think it is better if the back-end can give the front-end the ID of the default timeline.

Reply:
I have since added a `defaultTimeline` attribute to the index response for timelines. `TimelinesStack` will be able to find the default time in O(1) time.